### DISCUSSION TITLE
Permission denied (publickey) when deploy from local machine to server

### DISCUSSION THREAD

**Discussion by svenbartak** [deployphp/deployer#2865] [Help needed]
Hi guys,

i have a strange problem with deploying an app to my remote server.

First of all, i have updated deployer from 6.x. to 7.x-dev, because of php8 deps. Before this upgrade the deployment works fine and smootly. Now i run into a repository permission denied (publickey) error if deploy to remote. Strangely, if i am using the same git clone command as deploy user on my remote machine it works fine.

**Log**
```
bin/dep deploy
deploy@dev.xxx.foo's password: *******
task deploy:check    
task deploy:info
[dev.xxx.foo] info deploying dev-deploy
task deploy:setup
task deploy:lock
task deploy:release
task deploy:update_code
[dev.xxx.foo]  error  in update_code.php on line 53:
[dev.xxx.foo] run [ -f /var/www/dev.xxx.foo/.dep/repo/HEAD ] || /usr/bin/git clone --mirror git@bitbucket.org:my-company/my-repo.git /var/www/dev.xxx.foo/.dep/repo/ 2>&1
[dev.xxx.foo] Cloning into bare repository '/var/www/dev.xxx.foo/.dep/repo/'...
[dev.xxx.foo] git@bitbucket.org: Permission denied (publickey).
[dev.xxx.foo] fatal: Could not read from remote repository.
[dev.xxx.foo] Please make sure you have the correct access rights
[dev.xxx.foo] and the repository exists.
[dev.xxx.foo] exit code 128 (Invalid exit argument)
task deploy:failed
task deploy:unlock
```

**Hints**
1. It exists an deploy user on the remote machine that have acces rights to the repository
2. If i connect manually to the remote machine and fire the git clone command, it works
3. With deployer 6 it works also before

It seems to be that deployer uses not the deploy user for cloning the repo.

**Configuration**
```
<?php

namespace Deployer;

require 'recipe/symfony.php';
require 'deployer/check.php';
require 'deployer/migrate.php';
require 'deployer/server.php';
require 'deployer/encore.php';

$projectName = 'dev.xxx.foo';
$projectPath = '/var/www/dev.xxx.foo';

$devUrls = 'dev.xxx.foo';
set('application', $projectName);
set('repository', 'git@bitbucket.org:my-company/my-repo.git');
set('shared_dirs', ['var/log', 'var/sessions', 'public/upload']);
set('shared_files', ['.env']);
set('writable_dirs', ['var']);
set('allow_anonymous_stats', false);
set('http_user', 'www-data');
set('http_group', 'www-data');
set('composer_action', 'install');
set('composer_options', '{{composer_action}} --verbose --prefer-dist --no-progress --no-interaction --no-dev --optimize-autoloader --apcu-autoloader');
set('console_options', '');

host($devUrls)
    ->setRemoteUser('deploy')
    ->setPort(22)
    ->setForwardAgent(true)
    ->setDeployPath($projectPath)
    ->set('branch', function () {
        return input()->getOption('branch') ?: 'dev-deploy';
    });

task('build', function () {
    run('cd {{release_path}} && build');
});

task('restart-fpm', function () {
    run('sudo /bin/systemctl restart php8.0-fpm');
});

before('deploy', 'deploy:check');
after('deploy:vendors', 'deploy:assets');
before('deploy:symlink', 'deploy:migrate');
after('deploy:symlink', 'restart-fpm');
after('deploy:failed', 'deploy:unlock');
```

I am out of ideas, hopefully some one can help me.

Thanks,
BR,
Sven

**Comment by antonmedv** [comment]
git@bitbucket.org: Permission denied (publickey).

**Reply by svenbartak** [reply]
Unfortunately I can't get anything going with this answer right now, what exactly do you mean by that?

**Reply by antonmedv** [reply]
Bitbucket says you key is incorrect. 

Try to use deploy keys/read only instead of agent forwarding.

**Comment by vaites** [comment]
I had this problem and I lost a lot of time with it. I created an SSH key (passwordless) on the server and added it to BitBucket, so I can clone the repo without password (tested via SSH). The deploy failed because **agent forwarding is enabled by default** so if your local machine can't connect o BitBucket the deploy will fail (that's my case). Just disable it with `set('forward_agent', false);`.
